### DISCUSSION TITLE
Reloading child BusinessListBase object inside a parent ReadOnlyBase object on database update

### DISCUSSION THREAD

**Discussion by CSLASFRMQ** [MarimerLLC/csla#3023] [Questions]
I'm very new to CSLA and have been roped in to fix bugs in a legacy code that uses CSLA. I'm facing the following issue,

Our microservice loads a ReadOnlyBase during the first call and I'm assuming it caches al the data in memory after that. Another microservice directly updates the database. How do I force refresh the child object? We do have a notification service that notifies changes to the database(I can know which is the corresponding CSLA object) but I don't know how to do this?

The following is a sample code,

```
public class C : BusinessBase<C>
    {
        public int Id { get; set; }
        public string Value { get; set; }
        public bool IsActive { get; set; }
    }

    public class CList : BusinessListBase<CList, C>
    {
        
    }

    public class B : ReadOnlyBase<B>
    {
        public CList Cs
        {
            get;set;
        }
    }


    public class A : ReadOnlyBase<A>
    {
        public B Games
        {
            get; set;
        }
    }
```
1. Object of Class A is my main object, 
2. There is a property of class B as a child inside it.
3. Class B has a child property of a BusinessListBase CList. This is collection that is supposed to be refreshed.

Is this possible to do in CSLA?

**Comment by Chicagoan2016** [comment]
Your class hierarchy doesn't look right to me, a read/write collection inside a read-only object (Class B)?

**Reply by CSLASFRMQ** [reply]
I cannot copy paste my actual code here. May be my sample code is wrong. But can you understand the issue I'm facing?

**Reply by Chicagoan2016** [reply]
In your collection object C you could have a static method, say call it 'InvalidateCache' and inside that method you would clear the cache, you could call 'InvalidateCache' from your notification service, that way your collection C object will get the 'fresh' data from database next time you call it's get method.

**Reply by CSLASFRMQ** [reply]
Our notification service sends me the FullName of the type that was changed. So I'm getting the change as {my namespace}.B. I'm invalidating the cache in B. My problem is my code references the object A and object A has a property of type B as a child which still loads from the memory and has stale data.

**Reply by rockfordlhotka** [reply]
It depends on how the value is cached. Is it in a static field, or REDIS, or ???.

It is pretty common for the caching code to be in the static factory method in your class, though that's not a guarantee.

**Reply by CSLASFRMQ** [reply]
I'm not clear on how the caching works in CSLA. The actual business object corresponding to the table is cached and once the cache gets cleared it has the latest data. But other business objects that has the changed business object as a child never gets refreshed. I'm sorry if I keep repeating the same thing again.

**Reply by Chicagoan2016** [reply]
If you are getting stale data in object A (for object B) then most probably you are caching object B inside object A.
You will have to clear that 'cache' as well.

**Reply by rockfordlhotka** [reply]
CSLA doesn't handle caching. Any caching is handled in the application code itself. If the read-only type has a static factory method (like `GetObjectB`) then look in that code to see how it is caching the value.
